### ISSUE TITLE
Assign "void" from a function array_diff().

### ISSUE THREAD

**Issue by tanzanite777** [DEVSENSE/phptools-docs#573]
Type: <b>Bug</b>

Weird bug that i have not managed to find any cause for - nor any repro.

I have several such distracting false errors. Some of them disappear the moment you look at them (ie. open the file they are reported in for the first time since restarting vscode - i use "php.problems.scope": "all"). Some of them, however, never go away (or i do not know which file to open to trigger whatever solves the issue).

Typical example (note: illustrative, does not reproduce issue):
```
class Foo extends Bar {
    // note: getReegel does not exist anywhere else in the entire codebase (not in Bar nor anywhere else unrelated to Foo/Bar).
    // This, example case, function is used twice in the entire codebase - once in the same class (where the error does not go away)
    // and once in another unrelated class (where it goes away when looking at it).
    // Note that changing the function to explicitly return mixed ("public function getReegel() : mixed"):
    // * has no immediate effect - false errors remain
    // * however, restarting does make the false error disappear (currently using this workaround for all except array_diff)
    // * have not found a workaround for the error to disappear in array_diff case
    /**
     * @return mixed
     */
    public function getReegel(){
        return $this->reegel;
    }

/* snip */
    public function snafu() {
            $reegel = $this->getReegel(); // shows error here
    }
}
```
Note that array_diff (in the class that extends Foo) is also one of the persistent errors that never goes away (even when opening the devsense stub file for it).

I have no idea how to debug this issue - i can not share the codebase exhibiting the issue :( and have failed to find a reproduction so far. Random guess is that something does not get finished in static analysis - based on the fact that opening some files makes 7 out of 10 false errors disappear. Something akin to deadlock between source files that have references to each other.

This bug probably has been existed for eons and is not a regression - i have been cleaning up, the issue exhibiting astonishingly horrible, codebase (had ~1000+ issues) and this report covers the only detected issues left (ie. not anymore hidden by 1000+ other issues).

Extension version: 1.47.15512
VS Code version: Code 1.89.1 (dc96b837cf6bb4af9cd736aa3af08cf8279f7685, 2024-05-07T05:13:33.891Z)
OS version: Windows_NT x64 10.0.19045
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz (8 x 3600)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|undefined|
|Memory (System)|15.88GB (5.59GB free)|
|Process Argv|--crash-reporter-id adf79977-8c83-4303-918e-a7dfc5f31bf1|
|Screen Reader|no|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vspor879:30202332
vspor708:30202333
vspor363:30204092
vscod805cf:30301675
binariesv615:30325510
vsaa593:30376534
py29gd2263:31024239
c4g48928:30535728
azure-dev_surveyone:30548225
a9j8j154:30646983
962ge761:30959799
pythongtdpath:30769146
welcomedialog:30910333
pythonidxpt:30866567
pythonnoceb:30805159
asynctok:30898717
pythontestfixt:30902429
pythonregdiag2:30936856
pythonmypyd1:30879173
pythoncet0:30885854
h48ei257:31000450
pythontbext0:30879054
accentitlementsc:30995553
dsvsc016:30899300
dsvsc017:30899301
dsvsc018:30899302
cppperfnew:31000557
dsvsc020:30976470
pythonait:31006305
jchc7451:31067544
showvideot:31016892
chatpanelc:31048052
dsvsc021:30996838
bdiig495:31013172
pythoncenvpt:31062603
a69g1124:31058053
dvdeprecation:31061160
pythonprt:31056678
dwnewjupyter:31046869
fa862199:31048878

```

</details>

**Comment by jakubmisek** [comment]
Thank you,

- Do you have the class `Bar` somewhere? I don't have any error for this test case.
![image](https://github.com/DEVSENSE/phptools-docs/assets/842150/add6ba23-2c3c-4755-8a09-d01c772b7b84)
- What's the issue with `array_diff`?

**Comment by tanzanite777** [comment]
array_diff has the exact same issue ("Assigning "void" from a function array_diff() PHP(PHP1408)" for code "$diff = array_diff($list,$keys);")

... and i just managed to extract an reproduction of it (don't know what determines the order the files get analyzed - ie. there might be some file creation/modification date or filename dependencies or which file is opened in editor on restart for the issue to manifest itself). Tried to extract reprodution for getReegel to have the same problem and succeeded, for a moment, and then it disappeared when i tried to cut get_samm_seletus function down - reverting changes / restarting vscode did not bring it back though (which makes me believe even more that there is some race condition at play). Left the getReegel part in for illustrative purposes (as my previous non-repro illustration omitted that FooBar also declared $reegel as "private $reegel;" and Bar contained nothing of obvious relevance - no overloading at play etc).

Create new worskspace with the next four files:

File: class_api.php
```
class Class_API extends Main {

    public function __construct() {
    }

}
```

File: class_main.php
```
class Main {

    function __construct() {}

}
```

File: Import.class.php
```
class Import extends Class_API{

    private $reegel;

    public function __construct($iid=0){
    }

    /**
     * @return mixed
     */
    public function getReegel() {
        return $this->reegel;
    }

    /**
     * @param $samm
     * @return mixed
     */
    public function get_samm_seletus($samm){
        global $DB;

        $sammud = array();
        if($DB->something()){
        }else{
            $reegel = $this->getReegel(); // (1) Assigning "void" from a function getReegel() PHP(PHP1408)
            $res = $DB->query("SELECT * FROM impordireegel_samm WHERE impordireegel = $reegel ORDER BY jrk");

            foreach($res as $v){
                $sammud[$v['kood']] = $v;
            }

            $kokku = count($res);
        }

        return $sammud;
    }

}
```

File: ImportKontroll.class.php
```
class ImportKontroll extends Import {

    public function __construct($iid=0){
    }

    /**
     * @param $sailik
     * @return array
     */
    public function faili_orignimi_kontroll($sailik){
        global $DB;

        $failid = array();
        $res = $DB->query("whatever");

        $data = array();
        $koik_uus = array();
        foreach($res as $v){

            $eid = $v['id'];
            $asendused = $DB->something();

            $keys = array_keys($asendused);
            $koik_vana = array_merge($koik_vana,$keys);
            foreach($failid as $kaust=>$list){
                /** @trace $list */ // noticed that array_diff does not like $list and trace gives for me: "@trace $list: void" ... O_o
                $diff = array_diff($list,$keys); // (2) Assigning "void" from a function array_diff() PHP(PHP1408)
                $d['erinev'] = count($diff);
                $d['kokku']  = count($list);
                $d['kaal'] = ($d['erinev']*1.0)/$d['kokku'];

                $data[$eid][$kaust] = $d;
                $koik_uus = array_merge($koik_uus,$list);
            }
        }

        return $data;
    }
}
```

Reminder:
(1) Is purely illustrative - at least for me, It does not manifest in this sample code - however, without exceptions, manifests in real code (vs. several others that instantly disappear when i open the file issue report points at). 
(2) Is a reproduction - at least for me. It always manifests in given sample.

Nr 1 is hard to pin down (looks like some kind of race condition) - so, hopefully nr 2 is reproducible at your end.

**Comment by jakubmisek** [comment]
thanks!

(1) weird, I'll make more tests

(2) is caused by an invalid analysis rule found in standard phpstorm.meta.php:
```
override (\array_diff(0), type(0));
```

Will be fixed.

**Comment by jakubmisek** [comment]
(1) happens sometimes even for simple tests like:

```php
class Import {

    private $reegel;

    /**
     * @return mixed
     */
    public function zzz() {
        return $this->reegel;
    }

    public function get_samm_seletus($samm){
        $reegel = $this->zzz(); // (1) Assigning "void" from a function getReegel() PHP(PHP1408)
    }
}
```

This must be an old legacy behavior/bug ... working on it.

**Comment by jakubmisek** [comment]
In this case this is sort of right, but I agree not what we want.

`$this->reegel` is never assigned to, so it is treated as uninitialized, and the editor just did not "believe" the `@return mixed` annotation because it's too generic :-)

I'll clean this behavior a bit, it should be alright next update.

**Comment by jakubmisek** [comment]
both issues fixed in the latest update 1.48

thanks!
